src/charm.py

```diff
@@ -1,34 +1,146 @@
+
+    def _on_config_changed(self, _: ops.charm.ConfigChangedEvent) -> None:
+        """Event handler for configuration changed events.
+        Args:
+            _ (ops.charm.ConfigChangedEvent): The configuration changed event
+        Returns:
+            None: None
+        """
+        # TODO
+        # - update existing mongo configurations based on user preferences
+        # - add additional configurations as according to spec doc
+        with open("/etc/mongod.conf") as mongo_config_file:
+            mongo_config = yaml.safe_load(mongo_config_file)
+
+        machineIP = str(self.model.get_binding(PEER).network.bind_address)
+        bindIPs = "localhost," + machineIP
+        mongo_config["net"]["bindIp"] = bindIPs
+        if "replication" not in mongo_config:
+            mongo_config["replication"] = {}
+
+        mongo_config["replication"]["replSetName"] = "rs0"
+        with open("/etc/mongod.conf", "w") as mongo_config_file:
+            yaml.dump(mongo_config, mongo_config_file)
+
+        self.unit.status = ops.model.ActiveStatus()
+
+    def _on_start(self, event: ops.charm.StartEvent) -> None:
+        """Enables MongoDB service and initializes replica set
+        Args:
+            event (ops.charm.StartEvent): The triggering start event 
+        Returns:
+            None: None
+        """
+        logger.debug("Running on_start")
+        if not self.unit.is_leader():
+            return
+
+        # start mongo service
+        self._open_port_tcp(self.port)
+        self.unit.status = MaintenanceStatus("enabling MongoDB")
+        try:
+            logger.debug("enabling mongodb")
+            service_resume("mongod.service")
+            service_restart("mongod.service")
+        except subprocess.CalledProcessError as e:
+            logger.error("failed to enable mongo error: %s", e)
+            self.unit.status = BlockedStatus("failed to enable mongo")
+            return
+
+        if not self.mongo.is_ready():
+            self.unit.status = WaitingStatus("Waiting for MongoDB Service")
+            event.defer()
+            return
 
-    def _add_mongodb_org_repository(self):
+        # initialize replica set
+        self.unit.status = MaintenanceStatus("initializing MongoDB replicaset")
+        logger.debug("initalizing replica set")
+        try:
+            logger.debug("initialzing replica set for these IPs %s", self.unit_ips)
+            self.mongo.initialize_replica_set(self.unit_ips)
+            self.peers.data[self.app]["replica_set_hosts"] = json.dumps(self.unit_ips)
+        except Exception as e:
+            logger.error("Error initializing replica sets in _on_start: error={}".format(e))
+            self.unit.status = BlockedStatus("failed to initialize replicasets")
+            return
+
+        self.unit.status = ActiveStatus("MongoDB started")
+
+    def _open_port_tcp(self, port: int) -> None:
+        """Open the given port.
+        Args:
+            port (int): The port to open
```

Comment:
```suggestion
            port: The port to open
```

No need to include this when you're using type annotations so broadly :)